### PR TITLE
[Bounty] Faster-RCNN bring up using TTNN APIs (#29359)

### LINKED ISSUE #29359: [Bounty $1500] Faster-RCNN bring up using TTNN APIs
# Faster-RCNN bring up using TTNN APIs

## 📝 Background

This bounty is for bringing up the [Faster-RCNN](https://docs.pytorch.org/vision/main/models/faster_rcnn.html) model using TTNN APIs on Tenstorrent hardware (Wormhole or Blackhole). Faster-RCNN is a widely used deep learning model for object detection in images, applied in autonomous vehicles, surveillance, robotics, and various vision-based industrial applications. The objective is to enable the model to run efficiently on TT hardware, allowing Tenstorrent users and partners to leverage high-throughput, low-latency inference for object detection tasks.

## 🎯 What Success Looks Like

A successful submission will fulfill all requirements in the following stages. Payout is made after all three stages are completed.

### Stage 1 - Bring Up

- Implement Faster-RCNN using TTNN APIs (Python).
- Model runs on either N150 or N300 hardware with no errors.
- Produces valid object detections on sample images.
- Output is verifiable (e.g., bounding boxes, class labels).
- Achieves baseline throughput target (at least 10 FPS) as per bounty.
- Accuracy evaluated on COCO dataset following [Yolov4 demo instructions](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/yolov4#coco-2017-dataset).
- Clear instructions for setup and running the model.

### Stage 2 - Basic Optimizations

- Use optimal sharded/interleaved memory configs and sharding strategy for conv ops.
- Fuse simple ops where possible (e.g., relu+conv using TT fused ops).
- Store intermediates in L1 where beneficial.
- Use recommended TTNN/tt-metal CNN flows.
- Use TT library of fused ops like [relu with conv](https://github.com/tenstorrent/tt-metal/pull/29236)

### Stage 3 - Deeper Optimization

- Maximize core counts used per inference.
- Implement deeper TT-specific optimizations (e.g., grouped convolutions, reordering to avoid unnecessary Tensor manipulations).
- Minimize memory and TM overheads.
- Document any advanced tuning or known issues.

## 🧭 Guidance & Starting Points

- Use the [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md) as your primary reference.
- Reference [Yolov4 implementation (PR 29157)](https://github.com/tenstorrent/tt-metal/pull/29157) for example end-to-end flows.
- Use the [alternative PyTorch implementation](https://github.com/jwyang/faster-rcnn.pytorch) for model structure details.
- Refer to [TT fused ops](https://github.com/tenstorrent/tt-metal/pull/29236) for opportunities to optimize.
- Target input resolutions: start with 320x320x3, then 640x640x3.
- Ask for help or file issues if ops are missing in TTNN.

## 🔎 Possible Approaches

- Start from an existing PyTorch Faster-RCNN model and port layers one by one to TTNN.
- Validate each submodule's output against CPU/PyTorch reference before full integration.
- Experiment with different sharding strategies and memory configs for convolutions.
- Use TTNN profiling tools to identify bottlenecks and areas for fusion.
- Open a draft PR early to get feedback on your approach.

## 📊 Result submission guidelines notes

Beyond the model implementation itself. Contributors must submit the following material as a proof of work. However, feel free to open a PR at any time if you want us checking you are on the right track. Just understand that payout is only made after all 3 stages are completed.

* Functional model implementation
* Validation logs (output correctness)
* Performance report + header for final review

Links:
- [Link to perf sheet](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md#41-performance-sheet)
- [Perf header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

---

## 📚 Resources

- [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md)
- [CNN Bring-up & Optimization in TT-NN](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/CNNs/cnn_optimizations.md)
- [Yolov4 TTNN implementation PR](https://github.com/tenstorrent/tt-metal/pull/29157)
- [Faster-RCNN PyTorch implementation](https://github.com/jwyang/faster-rcnn.pytorch)
- [Faster-RCNN official docs](https://docs.pytorch.org/vision/main/models/faster_rcnn.html)
- [TT fused ops PR](https://github.com/tenstorrent/tt-metal/pull/29236)
- [Perf report header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

### PR DESCRIPTION
**Pull request by aliraza556**
### Ticket

Resolves https://github.com/tenstorrent/tt-metal/issues/29359

### Problem description

There is currently no Faster-RCNN implementation using TTNN APIs for Tenstorrent hardware.
Faster-RCNN is a widely used object detection model combining a Region Proposal Network (RPN)
with a Fast R-CNN detector, and its availability on TT hardware is needed to expand the
model coverage for vision workloads. This PR addresses the $1500 bounty to bring up
Faster-RCNN with ResNet-50-FPN backbone, targeting N150/N300 devices, with the goal of
producing valid COCO detections at >= 10 FPS and incorporating Stage 1 + Stage 2 optimizations.

### What's changed

Implemented Faster-RCNN with ResNet-50-FPN backbone using TTNN APIs. The computationally
intensive components — ResNet-50 backbone (50 conv layers across 16 bottleneck blocks),
Feature Pyramid Network (8 conv layers), and RPN head convolutions — run on the TT device.
Dynamic, data-dependent operations (anchor generation, NMS, ROI Align, box head FC layers)
fall back to CPU via torchvision.

**Stage 1 - Bring Up:**
- Full end-to-end Faster-RCNN inference pipeline producing valid object detections
- Pretrained COCO weights from torchvision (91 classes)
- PCC validation tests against PyTorch reference (threshold >= 0.90)
- Demo with sample image inference and performance benchmarking
- Target input resolution: 320x320

**Stage 2 - Basic Optimizations:**
- All `FrozenBatchNorm2d` layers folded into preceding `Conv2d` weights/biases
- ReLU fused with `Conv2d` via `ttnn.UnaryWithParam(ttnn.UnaryOpType.RELU)`
- `bfloat8_b` weights with `MathFidelity.LoFi` for max throughput
- `HEIGHT_SHARDED` memory layout for conv operations
- Weights double buffering enabled
- Intermediate tensors explicitly deallocated to minimize memory footprint

**New Files:**

| File | Description |
|------|-------------|
| `models/demos/vision/detection/faster_rcnn/common.py` | Constants, COCO class names, model loader with configurable transform |
| `models/demos/vision/detection/faster_rcnn/reference/faster_rcnn.py` | PyTorch reference wrapper for validation |
| `models/demos/vision/detection/faster_rcnn/tt/model_preprocessing.py` | Weight extraction, FrozenBatchNorm2d folding, TTNN tensor conversion |
| `models/demos/vision/detection/faster_rcnn/tt/ttnn_resnet50_backbone.py` | ResNet-50 backbone in TTNN (16 bottleneck blocks, 50 conv layers) |
| `models/demos/vision/detection/faster_rcnn/tt/ttnn_fpn.py` | Feature Pyramid Network in TTNN (4 inner + 4 layer convolutions) |
| `models/demos/vision/detection/faster_rcnn/tt/ttnn_faster_rcnn.py` | Full model: TtFasterRCNN, TtRPNHead, TtFasterRCNNBackboneOnly |
| `models/demos/vision/detection/faster_rcnn/tests/pcc/test_faster_rcnn.py` | PCC validation: backbone, full model output, output format |
| `models/demos/vision/detection/faster_rcnn/demo/demo.py` | Demo inference + performance benchmark |
| `models/demos/vision/detection/faster_rcnn/README.md` | Setup and usage documentation |

**Architecture:**

```
Input (320x320x3)
    |
    v
+---------------------+
|  ResNet-50 Backbone  |  <- TTNN (conv1 + 16 bottleneck blocks)
|  (C2, C3, C4, C5)   |
+---------------------+
    |
    v
+---------------------+
|        FPN           |  <- TTNN (inner 1x1 + layer 3x3 convs)
|  (P2, P3, P4, P5,   |
|   pool)              |
+---------------------+
    |
    v
+---------------------+
|     RPN Head         |  <- TTNN (3x3 conv + 1x1 cls/bbox)
|  (objectness + bbox) |
+---------------------+
    |
    v
+---------------------+
|  Anchor Gen + NMS    |  <- CPU (torchvision)
|  ROI Align           |
|  Box Head + Predict  |
+---------------------+
    |
    v
  Detections (boxes, labels, scores)
```

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:aliraza/faster-rcnn-ttnn-bringup)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:aliraza/faster-rcnn-ttnn-bringup)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:aliraza/faster-rcnn-ttnn-bringup)
- [ ] New/Existing tests provide coverage for changes

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:aliraza/faster-rcnn-ttnn-bringup)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:aliraza/faster-rcnn-ttnn-bringup)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:aliraza/faster-rcnn-ttnn-bringup)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212806203871186

### CODE REVIEW

**Comment by aliraza556** [comment]
@jberkowitzTT @uaydonat @yieldthought

 **CI Note:** The `t3000-apc-fast-tests / T3000 fast tests` failure is **unrelated to this PR**. The failing test is `test_tt_fabric` (C++ fabric routing perf microbenchmark), which failed because 2 `NeighborExchangeUnicast` bandwidth tests exceeded the 5% golden comparison tolerance — due to performance being *better* than the golden baseline (~6-9% faster), not a regression. This PR only modifies Python model files under `models/demos/vision/detection/faster_rcnn/` and has no connection to the fabric routing test infrastructure.

image

**Comment by aliraza556** [comment]
@jberkowitzTT The branch is now fully rebased and up to date with `main`. Could you please re-run the CI workflows when you get a chance? Thanks!

**Review by yieldthought — COMMENTED** [review]
Code looks good, what are the performance and accuracy figures? Can you include test results to demonstrate that this runs and works please?

File: models/demos/vision/detection/faster_rcnn/common.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            min_size=min_size,
+            max_size=max_size,
+        )
+    else:
+        model = fasterrcnn_resnet50_fpn(
+            weights=None,
+            num_classes=FASTER_RCNN_NUM_CLASSES,
+            min_size=min_size,
+            max_size=max_size,
+        )
```
Comment (yieldthought) [inline models/demos/vision/detection/faster_rcnn/common.py:127]: `pretrained=False` still keeps the default pretrained ResNet-50 backbone in torchvision unless `weights_backbone=None` is passed explicitly. That makes this helper return a different model than its flag suggests.

```suggestion
        model = fasterrcnn_resnet50_fpn(
            weights=None,
            weights_backbone=None,
            num_classes=FASTER_RCNN_NUM_CLASSES,
            min_size=min_size,
            max_size=max_size,
        )
```

**Review by yieldthought — APPROVED** [review]

**Comment by aliraza556** [comment]
> Perfect, I'll tag in the bounty author, @dvartaniansTT for the final review + approval

@jberkowitzTT Cool
